graph/src/util/intern.rs

```diff
@@ -0,0 +1,373 @@
+//! Interning of strings.
+//!
+//! This module provides an interned string pool `AtomPool` and a map-like
+//! data structore `Object` that uses the string pool. It offers two
+//! different kinds of atom: a plain `Atom` (an integer) and a `FatAtom` (a
+//! reference to the pool and an integer). The former is useful when the
+//! pool is known from context whereas the latter carries a reference to the
+//! pool and can be used anywhere.
+
+use std::convert::TryFrom;
+use std::{collections::HashMap, sync::Arc};
+
+use super::cache_weight::CacheWeight;
+
+// We could probably get away with a `u16` here, but unless we improve the
+// layout of `Object`, there's little point in that
+type AtomInt = u32;
+
+/// An atom in a pool. To look up the underlying string, surrounding code
+/// needs to know the pool for it.
+#[derive(Eq, Hash, PartialEq, Clone, Copy, Debug)]
+pub struct Atom(AtomInt);
+
+/// An atom and the underlying pool. A `FatAtom` can be used in place of a
+/// `String` or `Word`
+pub struct FatAtom {
+    pool: Arc<AtomPool>,
+    atom: Atom,
+}
+
+impl FatAtom {
+    pub fn as_str(&self) -> &str {
+        self.pool.get(self.atom).expect("atom is in the pool")
+    }
+}
+
+impl AsRef<str> for FatAtom {
+    fn as_ref(&self) -> &str {
+        self.as_str()
+    }
+}
+
+pub enum Error {
+    NotInterned(String),
+}
+
+/// A pool of interned strings. Pools can be organized hierarchically with
+/// lookups in child pools also considering the parent pool. The chain of
+/// pools from a pool through all its ancestors act as one big pool to the
+/// outside.
+pub struct AtomPool {
+    base: Option<Arc<AtomPool>>,
+    base_sym: AtomInt,
+    atoms: Vec<Box<str>>,
```

Comment:
If you are attempting to reduce memory consumption, would it be better to use `Arc` instead of `Box` here? The `str` in `atoms` and `words` are identical.

Reply:
Yes, that might be a win depending on the average size of those strings. `Arc` introduces an overhead of 24 bytes, whereas `str` is 16 bytes, so the savings come down to how many strings fit into 8 bytes. But for now, the main win of interning will be to reduce hundreds of copies of the same string to two.